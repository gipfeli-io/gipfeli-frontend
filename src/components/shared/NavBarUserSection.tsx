import {Button, Stack} from '@mui/material'
import React from 'react'
import {signIn, signOut} from 'next-auth/react'
import {useAuth} from '../../hooks/use-auth'
import Typography from '@mui/material/Typography'

function NavBarUserSection() {
    const { user, isAuthenticated } = useAuth();

    if (!isAuthenticated) {
        return (
            <Stack spacing={2} direction={'row'}>
                <Button variant={'contained'} id={'join-button'}>Join</Button>
                <Button variant={'outlined'} onClick={() => signIn()}>Login</Button>
            </Stack>
        )
    }

    return (
        <Stack spacing={2} direction={'row'}>
            <Typography variant="h6" component="div" sx={{flexGrow: 1}}>
                Hello, {user}!
            </Typography>
            <Button variant={'outlined'} onClick={() => signOut()} id={'logout-button'}>Logout</Button>
        </Stack>
    )

}

export default NavBarUserSection