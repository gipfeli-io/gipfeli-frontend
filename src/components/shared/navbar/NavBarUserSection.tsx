import { Button, Stack } from '@mui/material'
import React from 'react'
import Typography from '@mui/material/Typography'
import { Link } from 'react-router-dom'
import useAuth from '../../../hooks/use-auth'
import { useNavigate } from 'react-router'

function NavBarUserSection () {
  const auth = useAuth()
  const navigate = useNavigate()

  if (!auth.username) {
    return (
      <Stack spacing={2} direction={'row'}>
        <Button variant={'contained'} id={'join-button'}>Join</Button>
        <Button component={Link} to='login' variant={'outlined'} color='inherit'>Login</Button>
      </Stack>
    )
  }

  return (
    <Stack spacing={2} direction={'row'}>
      <Typography variant="h6" component="div" sx={{ flexGrow: 1 }}>
        Hello, {auth.username}!
      </Typography>
      <Button variant={'outlined'} color='inherit' id={'logout-button'}
              onClick={() => auth.signOut(() => navigate('/login'))}>Logout</Button>
    </Stack>
  )
}

export default NavBarUserSection
