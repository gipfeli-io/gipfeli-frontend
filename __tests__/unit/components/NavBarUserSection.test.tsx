import '@testing-library/jest-dom'
// @ts-ignore
import React from 'react'
// @ts-ignore
import renderer from 'react-test-renderer'
import NavBarUserSection from '../../../src/components/shared/NavBarUserSection'
import { MemoryRouter } from 'react-router'
import { render } from '@testing-library/react'

const mockSignIn: jest.Mock = jest.fn()
const mockSignOut: jest.Mock = jest.fn()
const mockAuthenticationContext: any = {
  username: null,
  token: null,
  mockSignIn,
  mockSignOut
}

const mockUseLocationValue = {
  pathname: '/localhost:3000/login',
  search: '',
  hash: '',
  state: null
}

const mockUsername = 'test@gipfeli.io'
const mockToken = 'mockedToken'

jest.mock('../../../src/hooks/use-auth', () => jest.fn().mockImplementation(() => mockAuthenticationContext))

jest.mock('react-router', () => ({
  ...jest.requireActual('react-router') as any,
  useNavigate: () => jest.fn(),
  useLocation: () => jest.fn().mockImplementation(() => mockUseLocationValue),
  useHref: () => jest.fn()
}))

describe('NavBarUserSection', () => {
  describe('test with logged in user', () => {
    beforeEach(() => {
      mockAuthenticationContext.token = mockToken
      mockAuthenticationContext.username = mockUsername
    })

    it('behaves consistently when logged in', () => {
      const tree = renderer
        .create(<MemoryRouter initialEntries={['/currentUri']}><NavBarUserSection/></MemoryRouter>)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it('shows logout button and name of user if logged in', async () => {
      mockAuthenticationContext.username = mockUsername
      mockAuthenticationContext.token = mockToken
      const { container, queryByText } = render(
    <MemoryRouter initialEntries={['/currentUri']}><NavBarUserSection/></MemoryRouter>
      )

      const logoutButton = container.querySelector('#logout-button')
      const joinButton = container.querySelector('#join-button')
      // @ts-ignore
      const greeting = queryByText('test@gipfeli.io', { exact: false })

      expect(logoutButton).toBeInTheDocument()
      expect(joinButton).not.toBeInTheDocument()
      expect(greeting).toBeInTheDocument()
    })
  })

  describe('test with logged out user', () => {
    beforeEach(() => {
      mockAuthenticationContext.token = null
      mockAuthenticationContext.username = null
    })

    it('behaves consistently when not logged in', () => {
      const tree = renderer
        .create(<MemoryRouter initialEntries={['/currentUri']}><NavBarUserSection/></MemoryRouter>)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('shows login button if not logged in', async () => {
      const { container } = render(
        <MemoryRouter initialEntries={['/currentUri']}><NavBarUserSection/></MemoryRouter>
      )

      const joinButton = container.querySelector('#join-button')
      const logoutButton = container.querySelector('#logout-button')

      expect(joinButton).toBeInTheDocument()
      expect(logoutButton).not.toBeInTheDocument()
    })
  })
})
