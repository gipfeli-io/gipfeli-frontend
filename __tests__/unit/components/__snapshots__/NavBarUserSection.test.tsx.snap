// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBarUserSection test with logged in user behaves consistently when logged in 1`] = `
<div
  className="css-e53awj-MuiStack-root"
>
  <div
    className="MuiTypography-root MuiTypography-h6 css-1juivf6-MuiTypography-root"
  >
    Hello, 
    test@gipfeli.io
    !
  </div>
  <button
    className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    id="logout-button"
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    Logout
  </button>
</div>
`;

exports[`NavBarUserSection test with logged out user behaves consistently when not logged in 1`] = `
<div
  className="css-e53awj-MuiStack-root"
>
  <button
    className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    id="join-button"
    onBlur={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    Join
    <span
      className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
  <a
    href={[MockFunction]}
    onClick={[Function]}
  >
    <button
      className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Login
      <span
        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </a>
</div>
`;
